### PR TITLE
59 gnosis

### PR DESCRIPTION
**Pull request by stonecoldpat**
Incorporating Gnosis Safe into our meta-transaction library.

Contract changes:
- We have renamed the contract "Proxy" to "GnosisProxy" to avoid typescript conflicts. 
- We have added ChainID to the SALT of a Gnosis Safe.

We have basic unit tests for the contract to send transactions. 

We have created a GnosisForwarder.

Some awkward parts of the code:
- In decodeTx() we return "-1" for replayprotection.
- We don't use a lot of the fields in gnosis safe, but I would advise not changing the contract code.  
- There are others that I cannot remember.

### CODE REVIEW

File: src/ts/forwarders/gnosisSafeForwarder.ts [hunk 1]
```diff
@@ -120,86 +182,61 @@ export class GnosisSafeForwarder extends WalletForwarder {
...
+    };
 
-    return defaultAbiCoder.encode(
-      ["uint", "address", "uint", "bytes"],
-      [defaulted.callType, defaulted.to, defaulted.value, defaulted.data]
-    );
+    return await this.signAndEncodeMetaTransaction(callData);
   }
 
   // Overrides the encodeAndSignParams from the Forwarder class.
```
Comment (stonecoldpat) [inline src/ts/forwarders/gnosisSafeForwarder.ts]: still got this comment

File: test/contracts/relayHub.test.ts [hunk 2]
```diff
@@ -317,6 +317,58 @@ describe("RelayHub Contract", () => {
...
+        replayProtectionAuthority,
+        forwarderAddress,
+        ChainID.MAINNET,
+      ]
+    );
+
+    return await owner.signMessage(arrayify(keccak256(encodedMetaTx)));
+  };
+
+  const encodeAndsignBatch = async (
```
Comment (stonecoldpat) [inline test/contracts/relayHub.test.ts]: capital S

File: src/ts/forwarders/gnosisSafeForwarder.ts (hunk 1 above)
Comment (yahgwai) [reply src/ts/forwarders/gnosisSafeForwarder.ts]: fixed

File: test/contracts/relayHub.test.ts (hunk 2 above)
Comment (yahgwai) [reply test/contracts/relayHub.test.ts]: fixed

File: README.md [hunk 3]
```diff
@@ -5,311 +5,40 @@ Ethereum transaction's intertwine the identity of who paid for the transaction (
...
 
-- **A candidate RelayHub standard**: Our RelayHub.sol is a minimal and potential candidate for the \_msgSender() standard.
-- **Minimal proxy contracts**: Our proxy contract is only responsible for forwarding and batching calls. It implements EIP-1167 to minimise storage/deployment overhead.
-- **Flexible replay protection:** We have implemented Nonce, MultiNonce and BitFlip, so you can decide if all transactions must be ordered, if you want up to N concurrent transactions, or to always support out-of-order transactions.
+- **A Universal Forwarder**: Our RelayHub.sol can be used for the \_msgSender() standard.
+- **Minimal wallet contract**: Our proxy contract only requires 67k gas to deploy & 26k gas per transaction. It is minimal code and supports batching transactions. As well, its replay protection supports out of order-transactions (bitflip) and concurrent transactions (multinonce).
+- **GnosisSafe**: We have incorporated GnosisSafe and our library tracks the replay protection nonce such that it is meta-transaction friendly. It is an audited wallet contract that is increasingly widely used.
```
Comment (yahgwai) [inline README.md]: Include hyperlink to gnosis safe

File: README.md [hunk 4]
```diff
@@ -5,311 +5,40 @@ Ethereum transaction's intertwine the identity of who paid for the transaction (
...
 
-Our repository is a protocol and relay-independent approach that any project can adopt. We hope it will make it easier for projects to tap into third party relayer APIs and to avoid re-implementing the wheel for reliable transaction infrastructure.
+**Our repository is a protocol and relay-independent approach** that any project can adopt. We hope it will make it easier for projects to tap into third party relayer APIs and to avoid re-implementing the wheel for reliable transaction infrastructure.
 
 # Getting started
 
-We assume you have set up your nodejs environment and you simply need to plug-in our library.
+We have put together a guide for the universial forwarder, proxy account and gnosis safe:
```
Comment (yahgwai) [inline README.md]: typo - universial

File: gnosisSafe.md [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+```
+
+The Gnosis Safe library only supports NONCE replay protection. Check out our [wallet contracts](proxyAccounts) for more exotic replay protection.
+
+3. Let's create the meta-transaction library with your preferred options!
+
+To instantiate the Gnosis Safe forwarder:
+
+```js
+const signer = Wallet.Mnemonic("");
```
Comment (yahgwai) [inline gnosisSafe.md]: should be:
```
const signer = Wallet.Mnemonic("").connect(provider);
```
maybe check all other places where we new up a wallet

File: gnosisSafe.md [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+  {
+    to: echoCon.address,
+    value: 0,
+    data: data,
+    revertOnFail: false,
+  },
+];
+```
+
+An additional feature is `revertOnFail` which lets you decide if the entire batch of transactions should revert if the meta-transaction fails. Again, we omit `CallType` as it should only be used by advanced users and 99% of meta-transactions only require the `.call` functionality.
```
Comment (yahgwai) [inline gnosisSafe.md]: Should remove 99 and put "often" or "most" :smile:

File: README.md (hunk 3 above)
Comment (stonecoldpat) [reply README.md]: fixed

File: README.md (hunk 4 above)
Comment (stonecoldpat) [reply README.md]: fixed

File: gnosisSafe.md (hunk 5 above)
Comment (stonecoldpat) [reply gnosisSafe.md]: fixed

File: gnosisSafe.md (hunk 6 above)
Comment (stonecoldpat) [reply gnosisSafe.md]: fixed

**Review by yahgwai — APPROVED** [review]

**Pull request merged**